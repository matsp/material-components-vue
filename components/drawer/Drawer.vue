<template>
  <aside
    :class="classes"
    class="mdc-drawer"
    @MDCDrawer:closed="onClosed"
  >
    <slot name="header" />
    <slot />
  </aside>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '../base'
import { MDCDrawer } from '@material/drawer'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'open',
    event: 'change'
  },
  props: {
    modal: {
      type: Boolean,
      default: false
    },
    dismissible: {
      type: Boolean,
      default: false
    },
    open: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      mdcDrawer: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-drawer--dismissible': this.dismissible,
        'mdc-drawer--modal': this.modal
      }
    },
    model: {
      get () {
        return this.open
      },
      set (value) {
        this.$emit('change', value)
      }
    }
  },
  watch: {
    modal () {
      if (!this.mdcDrawer) this.mdcDrawer = new MDCDrawer(this.$el)
    },
    dismissible () {
      if (!this.mdcDrawer) this.mdcDrawer = new MDCDrawer(this.$el)
    },
    open () {
      this.mdcDrawer.open = this.open
    }
  },
  mounted () {
    if (!this.mdcDrawer && (this.dismissible || this.modal)) { this.mdcDrawer = new MDCDrawer(this.$el) }
  },
  provide: {
    mdcDrawer: this
  },
  beforeDestroy () {
    if (this.mdcDrawer) this.mdcDrawer.destroy()
  },
  methods: {
    onClosed () {
      this.model = false
      this.$emit('closed')
    }
  }
}
</script>
