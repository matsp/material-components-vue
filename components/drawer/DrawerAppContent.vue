<template>
    <div class="mdc-drawer-app-content">
    <slot/>
  </div>
</template>

<script>
  import { baseComponentMixin, themeClassMixin } from '../base'

  export default {
  mixins: [baseComponentMixin, themeClassMixin]
}
</script>

<style scoped>

</style>
