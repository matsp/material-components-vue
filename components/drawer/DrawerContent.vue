<template>
  <div
    v-if="$slots.default"
    class="mdc-drawer__content"
    v-on="$listeners"
  >
    <slot />
  </div>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '@components/base'

export default {
  mixins: [baseComponentMixin, themeClassMixin]
}
</script>
