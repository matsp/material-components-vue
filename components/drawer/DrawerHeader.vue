<template>
  <div
    v-if="$slots['default'] || title || subTitle"
    class="mdc-drawer__header"
  >
    <h3
      v-if="title"
      class="mdc-drawer__title"
    >
      {{ title }}
    </h3>
    <h6
      v-if="subTitle"
      class="mdc-drawer__subtitle"
    >
      {{ subTitle }}
    </h6>
    <slot />
  </div>
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    title: {
      type: String,
      default: ''
    },
    subTitle: {
      type: String,
      default: ''
    }
  }
}
</script>
