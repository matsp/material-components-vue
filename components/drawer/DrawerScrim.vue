<template>
  <div class="mdc-drawer-scrim" />
</template>

<script>
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin]
}
</script>

<style scoped>

</style>
