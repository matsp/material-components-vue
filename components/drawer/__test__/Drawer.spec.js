import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import Drawer from '../Drawer.vue'
import DrawerAppContent from '../DrawerAppContent'
import DrawerHeader from '../DrawerHeader'
import DrawerContent from '../DrawerContent'
import DrawerScrim from '../DrawerScrim'

describe('Drawer', () => {
  it('should mount', () => {
    const wrapper = mount(Drawer)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(Drawer)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-drawer')
  })

  it('should render as modal', () => {
    const wrapper = mount(Drawer, {
      propsData: {
        modal: true
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-drawer--modal')
  })

  it('should render as dismissible', () => {
    const wrapper = mount(Drawer, {
      propsData: {
        dismissible: true
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-drawer--dismissible')
  })

  it('should render with title', () => {
    const wrapper = mount(Drawer, {
      propsData: {
        title: 'title'
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.mdc-drawer__header').exists()).toBe(true)
    expect(wrapper.find('h3').exists()).toBe(true)
    expect(wrapper.find('h3').classes()).toContain('mdc-drawer__title')
    expect(wrapper.find('h3').text()).toBe('title')
  })

  it('should render with subtitle', () => {
    const wrapper = mount(Drawer, {
      propsData: {
        subtitle: 'title'
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.mdc-drawer__header').exists()).toBe(true)
    expect(wrapper.find('h6').exists()).toBe(true)
    expect(wrapper.find('h6').classes()).toContain('mdc-drawer__subtitle')
    expect(wrapper.find('h6').text()).toBe('title')
  })
})

describe('DrawerAppContent', () => {
  it('should mount', () => {
    const wrapper = mount(DrawerAppContent)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(DrawerAppContent)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-drawer-app-content')
  })
})

describe('DrawerScrim', () => {
  it('should mount', () => {
    const wrapper = mount(DrawerScrim)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(DrawerScrim)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-drawer-scrim')
  })
})

describe('DrawerHeader', () => {
  it('should mount', () => {
    const wrapper = mount(DrawerHeader)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(DrawerHeader, {
      slots: {
        default: '123'
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-drawer__header')
    expect(wrapper.text()).toBe('123')
  })

  it('should render with title', () => {
    const wrapper = mount(DrawerHeader, {
      propsData: {
        title: 'title'
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('h3').exists()).toBe(true)
    expect(wrapper.find('h3').classes()).toContain('mdc-drawer__title')
    expect(wrapper.find('h3').text()).toBe('title')
  })

  it('should render with subtitle', () => {
    const wrapper = mount(DrawerHeader, {
      propsData: {
        subtitle: 'title'
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('h6').exists()).toBe(true)
    expect(wrapper.find('h6').classes()).toContain('mdc-drawer__subtitle')
    expect(wrapper.find('h6').text()).toBe('title')
  })
})

describe('DrawerContent', () => {
  it('should mount', () => {
    const wrapper = mount(DrawerContent)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(DrawerContent, {
      slots: {
        default: '123'
      }
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-drawer__content')
    expect(wrapper.text()).toBe('123')
  })
})
