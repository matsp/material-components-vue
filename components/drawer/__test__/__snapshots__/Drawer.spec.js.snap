// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer should render as dismissible 1`] = `
<aside class="mdc-drawer mdc-drawer--dismissible">
  <div class="mdc-drawer__header">
    <!---->
    <!---->
  </div>
</aside>
`;

exports[`Drawer should render as modal 1`] = `
<aside class="mdc-drawer mdc-drawer--modal">
  <div class="mdc-drawer__header">
    <!---->
    <!---->
  </div>
</aside>
`;

exports[`Drawer should render with no prop 1`] = `
<aside class="mdc-drawer">
  <div class="mdc-drawer__header">
    <!---->
    <!---->
  </div>
</aside>
`;

exports[`Drawer should render with subtitle 1`] = `
<aside class="mdc-drawer">
  <div class="mdc-drawer__header">
    <!---->
    <h6 class="mdc-drawer__subtitle">
      title
    </h6>
  </div>
</aside>
`;

exports[`Drawer should render with title 1`] = `
<aside class="mdc-drawer">
  <div class="mdc-drawer__header">
    <h3 class="mdc-drawer__title">
      title
    </h3>
    <!---->
  </div>
</aside>
`;

exports[`DrawerAppContent should render with no prop 1`] = `<div class="mdc-drawer-app-content"></div>`;

exports[`DrawerContent should render with no prop 1`] = `<div class="mdc-drawer__content">123</div>`;

exports[`DrawerHeader should render with no prop 1`] = `
<div class="mdc-drawer__header">
  <!---->
  <!----> 123</div>
`;

exports[`DrawerHeader should render with subtitle 1`] = `
<div class="mdc-drawer__header">
  <!---->
  <h6 class="mdc-drawer__subtitle">
    title
  </h6>
</div>
`;

exports[`DrawerHeader should render with title 1`] = `
<div class="mdc-drawer__header">
  <h3 class="mdc-drawer__title">
    title
  </h3>
  <!---->
</div>
`;

exports[`DrawerScrim should render with no prop 1`] = `<div class="mdc-drawer-scrim"></div>`;
